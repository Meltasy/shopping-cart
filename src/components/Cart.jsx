// const Cart = ({ buyProducts, addToCart }) => {
  const Cart = ({ productItems, totalBasket }) => {

  return (
    <div>
      <h1>Panier</h1>
      <p>{Object.entries(totalBasket)}</p>
      {/* {addToCart.map(i => (
        <p>{i.id} {i.quantity}</p>
      ))} */}
      {/* {buyProducts.map(item => {
        if (item.quantity !== 0) {
          <div key={item.id}>
            <h4>{item.title} â‚¬{item.price} {item.quantity}</h4>
            <img src={item.image} alt=' ' />
          </div>
        }
      })} */}
    </div>
  )
}
  
export default Cart
