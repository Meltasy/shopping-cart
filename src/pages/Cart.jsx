import CartItem from "../components/CartItem"

const Cart = ({ productItems, addToCart }) => {
  // const totalCostCart = ???

  return (
    <div>
      <h1>Panier</h1>
      <div className='cartItems'>
        {addToCart.map(item => (
          <CartItem
            key={item.id}
            itemId={item.id}
            quantity={item.quantity}
            productItems={productItems}
          />
        ))}
      </div>
      {/* <div>
        <h2>Total: â‚¬{totalCostCart}</h2>
      </div> */}
    </div>
  )
}
  
export default Cart
