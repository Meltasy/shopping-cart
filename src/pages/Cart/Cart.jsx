import CartItem from '../../components/CartItem/CartItem'

const Cart = ({ productItems, cartItems, totalCost, handleDelete }) => {

  return (
    <div>
      <h1>Cart</h1>
      <div className='cartItems'>
        {cartItems.map(item => (
          <CartItem
            key={item.id}
            itemId={item.id}
            price={item.price}
            quantity={item.quantity}
            productItems={productItems}
            handleDelete={handleDelete}
          />
        ))}
      </div>
      <div>
        <h2>Total: â‚¬{totalCost ? totalCost.toFixed(2): '0.00'}</h2>
      </div>
    </div>
  )
}
  
export default Cart
