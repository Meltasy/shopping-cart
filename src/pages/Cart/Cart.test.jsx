import { describe, it, expect, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event';
import Cart from './Cart'

const mockProductItems = [
  {
    id: '123',
    image: 'fakeImageA.jpg',
    title: 'Fake product A',
    price: 35.50,
    description: 'This is fake product A.'
  }, {
    id: '321',
    image: 'fakeImageB.jpg',
    title: 'Fake product B',
    price: 57.99,
    description: 'This is fake product B.'
  }, {
    id: '231',
    image: 'fakeImageC.jpg',
    title: 'Fake product C',
    price: 7.50,
    description: 'This is fake product C.'
  },
]
const mockCartItems = [
  { id: '123', price: 35.50, quantity: 3 },
  { id: '231', price: 7.50, quantity: 5 },
]

describe('Cart component', () => {
  beforeEach(() => {
    render(<Cart productItems={mockProductItems} cartItems={mockCartItems} />)
  })

  it('renders cart products', () => {
    const { container } = render(<Cart productItems={mockProductItems} cartItems={mockCartItems} />)
    expect(container).toMatchSnapshot()
  })

  it ('renders correct heading', () => {
    screen.debug()
    expect(screen.getByRole('heading', { level: 1 }).textContent).toMatch(/cart/i)
    expect(screen.getByRole('heading', { level: 1, name: 'Cart' })).toBeInTheDocument()
  })

  it('renders cart products with all complete details', () => {
    expect(screen.getByText('Fake product C')).toBeInTheDocument()
    expect(screen.getAllByRole('heading', { level: 4 })[2]).toHaveTextContent('€7.50')
    expect(screen.getAllByRole('heading', { level: 3 })[1]).toHaveTextContent('x 5')
    expect(screen.getAllByRole('heading', { level: 4 })[3]).toHaveTextContent('€37.50')
    expect(screen.getAllByRole('img')[1]).toHaveAttribute('src', 'fakeImageC.jpg')
  })

  it('renders correct number of cart products', () => {
    const cartItems = screen.getAllByRole('img')
    expect(cartItems).toHaveLength(mockCartItems.length)
  })
})

describe('Cart component calls totalCost and handleDelete as expected', () => {
  const totalCost = mockCartItems.reduce((total, item) => total + (item.price * item.quantity), 0)
  const mockDelete = vi.fn()

  beforeEach(() => {
    render(<Cart productItems={mockProductItems} cartItems={mockCartItems} totalCost={totalCost} handleDelete={mockDelete} />)
  })

  it('renders total cost of all cart products', () => {
    expect(screen.getByText(`Total: €${totalCost.toFixed(2)}`)).toBeInTheDocument()
  })
  
  it('calls handleDelete with correct arguments when button clicked', async () => {
    const user = userEvent.setup()
    const deleteBtn = screen.getAllByRole('button', { name: 'X'})
    await user.click(deleteBtn[0])
    expect(mockDelete).toHaveBeenCalledTimes(1)
    expect(mockDelete).toHaveBeenCalledWith('123', 35.50, 3)
  })
})
