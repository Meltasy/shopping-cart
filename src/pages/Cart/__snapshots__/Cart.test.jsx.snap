// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart component > renders cart products 1`] = `
<div>
  <div>
    <h1>
      Cart
    </h1>
    <div
      class="cartItems"
    >
      <div
        class="itemCart"
      >
        <div
          class="details"
        >
          <p>
            Fake product A
          </p>
          <h4>
            €
            35.50
          </h4>
          <h3>
            x 
            3
          </h3>
          <h4>
            €
            106.50
          </h4>
        </div>
        <img
          alt=" "
          src="fakeImageA.jpg"
        />
        <button>
          X
        </button>
      </div>
      <div
        class="itemCart"
      >
        <div
          class="details"
        >
          <p>
            Fake product C
          </p>
          <h4>
            €
            7.50
          </h4>
          <h3>
            x 
            5
          </h3>
          <h4>
            €
            37.50
          </h4>
        </div>
        <img
          alt=" "
          src="fakeImageC.jpg"
        />
        <button>
          X
        </button>
      </div>
    </div>
    <div>
      <h2>
        Total: €
        0.00
      </h2>
    </div>
  </div>
</div>
`;
